sql/analyzer/indexed_joins.go

```diff
@@ -858,117 +858,243 @@ func addMergeJoins(m *memo.Memo) error {
+		// Compute the max set of filter expressions that match that index
+		// While matchedFilters is not empty:
+		//    Check to see if any rIndexes match that set of filters
+		//    Remove the last matched filter
+
+		eqFilters := make([]filterAndPosition, 0, len(join.Filter))
+		for filterPos, filter := range join.Filter {
+			switch eq := filter.(type) {
 			case *memo.Equal:
-				l = f.Left
-				r = f.Right
-			default:
-				continue
-			}
+				l := eq.Left
+				r := eq.Right
 
-			if l.ScalarProps().Cols.Len() != 1 ||
-				r.ScalarProps().Cols.Len() != 1 {
-				continue
-			}
-
-			var swap bool
-			if l.ScalarProps().Tables.Contains(int(memo.TableIdForSource(leftGrp))) &&
-				r.ScalarProps().Tables.Contains(int(memo.TableIdForSource(rightGrp))) {
-			} else if r.ScalarProps().Tables.Contains(int(memo.TableIdForSource(leftGrp))) &&
-				l.ScalarProps().Tables.Contains(int(memo.TableIdForSource(rightGrp))) {
-				swap = true
-				l, r = r, l
-			} else {
-				continue
-			}
+				if l.ScalarProps().Cols.Len() != 1 ||
+					r.ScalarProps().Cols.Len() != 1 {
+					continue
+				}
 
-			lRef := getColumnRefFromScalar(l.Scalar)
-			rRef := getColumnRefFromScalar(r.Scalar)
+				// check that comparer is not non-decreasing
+				if !isWeaklyMonotonic(l.Scalar) || !isWeaklyMonotonic(r.Scalar) {
+					continue
+				}
 
-			// check that comparer is not non-decreasing
-			if !isWeaklyMonotonic(l.Scalar) || !isWeaklyMonotonic(r.Scalar) {
+				var swap bool
+				if l.ScalarProps().Tables.Contains(int(memo.TableIdForSource(leftGrp))) &&
+					r.ScalarProps().Tables.Contains(int(memo.TableIdForSource(rightGrp))) {
+				} else if r.ScalarProps().Tables.Contains(int(memo.TableIdForSource(leftGrp))) &&
+					l.ScalarProps().Tables.Contains(int(memo.TableIdForSource(rightGrp))) {
+					swap = true
+					l, r = r, l
+				} else {
+					continue
+				}
+				if swap {
+					eqFilters = append(eqFilters, filterAndPosition{&memo.Equal{
+						Left:  eq.Right,
+						Right: eq.Left,
+					}, filterPos})
+				} else {
+					eqFilters = append(eqFilters, filterAndPosition{eq, filterPos})
+				}
+			default:
 				continue
 			}
 
-			newFilters := make([]memo.ScalarExpr, len(join.Filter))
-			copy(newFilters, join.Filter)
-			// merge cond first
-			newFilters[0], newFilters[i] = newFilters[i], newFilters[0]
-
-			for _, rIdx := range sortedIndexScansForTableCol(rIndexes, rRef, join.Right.RelProps.FuncDeps().Constants(), rFilters) {
-				for _, lIdx := range sortedIndexScansForTableCol(lIndexes, lRef, join.Left.RelProps.FuncDeps().Constants(), lFilters) {
-					jb := join.Copy()
-					if d, ok := jb.Left.First.(*memo.Distinct); ok && lIdx.Idx.SqlIdx().IsUnique() {
-						jb.Left = d.Child
-					}
-					if d, ok := jb.Right.First.(*memo.Distinct); ok && rIdx.Idx.SqlIdx().IsUnique() {
-						jb.Right = d.Child
+		}
+		for _, lIndex := range lIndexes {
+			matchedEqFilters := matchedFiltersForLeftIndex(lIndex, join.Left.RelProps.FuncDeps().Constants(), eqFilters)
```

Comment:
functional dependencies are usually good at this -- the set of equality columns plus constants gives you an input ColSet. We've already built a ColSet for each index. We could use a combo of ColSet intersection and ordinal subtraction on the index ColSet to find all of the potential patching indexes (CRDB uses the phrase "InterestingOrderings" for the outcome of this process).